dds/DCPS/security/CryptoBuiltInImpl.cpp

```diff
@@ -1608,10 +1652,156 @@ bool CryptoBuiltInImpl::verify(const KeyMaterial& master, Session& sess,
+  char* afterSrtpsPrefix;
+  unsigned int sizeOfAuthenticated, sizeOfEncrypted;
+  char* encrypted = 0;
+
+  for (int i = 0; de_ser.available_r(); ++i) {
+    if (de_ser.available_r() < RTPS::SMHDR_SZ) {
+      return CommonUtilities::set_security_error(ex, -3, i, "Failed to deserialize SubmessageHeader");
+    }
+
+    ACE_CDR::Octet type, flags;
+    de_ser >> ACE_InputCDR::to_octet(type);
+    de_ser >> ACE_InputCDR::to_octet(flags);
+    de_ser.swap_bytes((flags & RTPS::FLAG_E) != ACE_CDR_BYTE_ORDER);
+    ACE_CDR::UShort octetsToNext;
+    de_ser >> octetsToNext;
+    de_ser.swap_bytes(Serializer::SWAP_BE);
+
+    if (i == 0 && type == RTPS::SRTPS_PREFIX) {
+      if (!(de_ser >> ch)) {
+        return CommonUtilities::set_security_error(ex, -4, i, "Failed to deserialize CryptoHeader");
+      }
+      if (octetsToNext < CRYPTO_HEADER_LENGTH || !de_ser.skip(octetsToNext - CRYPTO_HEADER_LENGTH)) {
+        return CommonUtilities::set_security_error(ex, -5, i, "Failed to find submessage after SRTPS_PREFIX");
+      }
+      afterSrtpsPrefix = mb_in.rd_ptr();
+      haveCryptoHeader = true;
+
+    } else if (haveCryptoHeader && type == RTPS::SEC_BODY) {
+      if (!(de_ser >> sizeOfEncrypted)) {
+        return CommonUtilities::set_security_error(ex, -13, i, "Failed to deserialize CryptoContent length");
+      }
+      const unsigned short sz = static_cast<unsigned short>(DCPS::max_marshaled_size_ulong());
+      if (sizeOfEncrypted + sz > octetsToNext) {
+        return CommonUtilities::set_security_error(ex, -14, i, "CryptoContent length out of bounds");
+      }
+      encrypted = mb_in.rd_ptr();
+      if (octetsToNext < sz || !de_ser.skip(octetsToNext - sz)) {
+        return CommonUtilities::set_security_error(ex, -15, i, "Failed to find submessage after SEC_BODY");
+      }
+
+    } else if (haveCryptoHeader && type == RTPS::SRTPS_POSTFIX) {
+      sizeOfAuthenticated = static_cast<unsigned int>(mb_in.rd_ptr() - afterSrtpsPrefix - RTPS::SMHDR_SZ);
+      const size_t availablePreFooter = de_ser.available_r();
+      if (!(de_ser >> cf)) {
+        return CommonUtilities::set_security_error(ex, -7, i, "Failed to deserialize CryptoFooter");
+      }
+      if (octetsToNext && availablePreFooter != octetsToNext) {
+        return CommonUtilities::set_security_error(ex, -8, i, "SRTPS_POSTFIX was not the final submessage");
+      }
+      haveCryptoFooter = true;
+      break;
+
+    } else if (octetsToNext == 0 && type != RTPS::PAD && type != RTPS::INFO_TS) {
+      break;
+
+    } else {
+      if (!de_ser.skip(octetsToNext)) {
+        return CommonUtilities::set_security_error(ex, -6, i, "Failed to find next submessage");
+      }
+    }
+  }
+
+  if (!haveCryptoHeader || !haveCryptoFooter) {
+    return CommonUtilities::set_security_error(ex, -9, 0, "Failed to find SRTPS_PREFIX/POSTFIX wrapper");
+  }
+
+  ACE_Guard<ACE_Thread_Mutex> guard(mutex_);
+  const KeySeq& keyseq = keys_[sending_participant_crypto];
```

Comment:
Do we need a check here first to see if the key exists in keys_ to avoid adding a default initialized one?